sodetlib/operations/bias_dets.py

```diff
@@ -201,3 +201,347 @@ def bias_to_rfrac(S, cfg, rfrac=0.5, bias_groups=None, iva=None,
 
     return biases
 
+def biasstep_rebias(
+        S, cfg, target_percentage_rn = 0.5, bias_groups=None,
+        math_only=False):
+
+    """
+    Scripts work well when detectors are already in transistion. For extreme cases
+    such as detectors are all in SC, it takes a few run to put them perfectly in
+    transition. Next step is making this script can handle extreme cases faster
+
+
+    """
+
+    ## take the initial biasstep
+    S.log("taking the first biasstep")
+    intical_current_biasvoltage = S.get_tes_bias_bipolar_array()
+    S.log(f"current biasvoltage {intical_current_biasvoltage}")
+    bsa_0 = bias_steps.take_bias_steps(S, cfg)
+    if not math_only:
+        bias_steps.plot_Rfrac(bsa_0)
+            
+    ## load IV analysis result so can use dynamic step size
+    iva = iv.IVAnalysis.load(bsa_0.meta['iv_file'])
+                
+    repeat_biasstep = False
+    ## check for script failue
+    for bl in bias_groups:
+        mask_bg = np.where(bsa_0.bgmap==bl)[0]
+        per_bl_R0 = bsa_0.R0[mask_bg]
+        if np.isnan(np.nanmedian(per_bl_R0)):
+            ## the yield is very low, could it be the biasstep analysis script failed? repeat again
+            repeat_biasstep = True
+            S.log("the biasstep script might have failed, going to repeat the measurement")
+
+    if repeat_biasstep:
+        S.log("retaking biasstep due to failure")
+        bsa_0 = bias_steps.take_bias_steps(S, cfg)
+        repeat_biasstep = False
+    
+    percentage_rn_0 = bsa_0.R0/bsa_0.R_n_IV
+    ## add in checks if detectors are SC  
+    bg_overbias_needed = []
+    overbias_needed = False
+    for bl in bias_groups:
+        mask_bg = np.where(bsa_0.bgmap==bl)[0]
+        per_bl_percentage_rn_0 = percentage_rn_0[mask_bg]
+        ## mask for SC
+        mask_sc = np.where((per_bl_percentage_rn_0 < 0.1) & (per_bl_percentage_rn_0 > -0.1))
+        ## if more than 10 percent of the detectors are SC
+        if len(mask_sc[0]) > 0.1*len(mask_bg):
+            bg_overbias_needed.append(bl)
+            overbias_needed = True
+
+    ## add in check if detectors are normal, hence we should increase the drop step
+    bg_detectors_normal = []
+    drop_from_normal = False
+    for bl in bias_groups:
+        mask_bg = np.where(bsa_0.bgmap==bl)[0]
+        per_bl_percentage_rn_0 = percentage_rn_0[mask_bg]
+        ## mask for normal
+        mask_normal = np.where((per_bl_percentage_rn_0 > 0.9) | (per_bl_percentage_rn_0 < -0.9))
```

Comment:
np.abs?

Reply:
I decided to keep it this way for now, this is for catching a bias-step failure as the current bias_step can return negative bias_resistance. The -0.9 threshold might be changed later with better understanding of the failure mode